# Step 2a: Data Cleaning, Pt. 2

In [1]:
# Import necessary libraries
import pandas as pd

## A. Grocery Market Data

In [2]:
usda_df = pd.read_csv('cleaned_usda_data.csv')

usda_df.head()

issuing_office report_date  program level_1 level_2 organic specialty unit  \
0  DES MOINES, IA  2024-08-30  POULTRY    EGGS     NaN      NO       NO   NaN   
1  DES MOINES, IA  2024-08-30  POULTRY    EGGS     NaN      NO       NO   NaN   
2  DES MOINES, IA  2024-08-30  POULTRY    EGGS     NaN      NO       NO   NaN   
3  DES MOINES, IA  2024-08-30  POULTRY    EGGS     NaN      NO       NO   NaN   
4  DES MOINES, IA  2024-08-30  POULTRY    EGGS     NaN      NO       NO   NaN   

           region  weighted_average_price  price_low  price_high  
0  NORTHEAST U.S.                     NaN        NaN         NaN  
1  NORTHEAST U.S.                     NaN        NaN         NaN  
2        NATIONAL                     NaN        NaN         NaN  
3        NATIONAL                     NaN        NaN         NaN  
4          ALASKA                     NaN        NaN         NaN

In [ ]:
# What's going on in level_2 re: NaN%?

nan_percentage = usda_df['level_2'].isna().sum() / len(usda_df['level_2']) * 100
print(nan_percentage)


8.058644049437694


### i. A deep dive on the organic column

In [ ]:
# What about in organic? 

nan_percentage = usda_df['organic'].isna().sum() / len(usda_df['organic']) * 100
print(nan_percentage)


56.39868954924173


In [5]:
# What all's in organic?

usda_df['organic'].unique()

array(['NO', 'YES', nan], dtype=object)

In [ ]:
# How much is 'YES' (i.e. organic)

organic_pct = (usda_df['organic'] == 'YES').sum() / len(usda_df['organic']) * 100
print(organic_pct)

3.4316235495295078


In [10]:
# Are the products marked 'YES' for organic labeled any differently in other columns?  What types of products are they?

organic_df = usda_df[usda_df['organic'] == 'YES']

organic_df.head()

issuing_office report_date  program  level_1 level_2 organic specialty  \
58  DES MOINES, IA  2024-08-30  POULTRY  CHICKEN   FRESH     YES       NO    
64  DES MOINES, IA  2024-08-30  POULTRY  CHICKEN   FRESH     YES       NO    
65  DES MOINES, IA  2024-08-30  POULTRY  CHICKEN   FRESH     YES       NO    
66  DES MOINES, IA  2024-08-30  POULTRY  CHICKEN   FRESH     YES       NO    
67  DES MOINES, IA  2024-08-30  POULTRY  CHICKEN   FRESH     YES       NO    

   unit          region  weighted_average_price  price_low  price_high  
58  NaN          ALASKA                     NaN        NaN         NaN  
64  NaN  SOUTHWEST U.S.                     NaN        NaN         NaN  
65  NaN  SOUTHWEST U.S.                     NaN        NaN         NaN  
66  NaN    MIDWEST U.S.                     NaN        NaN         NaN  
67  NaN    MIDWEST U.S.                     NaN        NaN         NaN

In [12]:
# Do any of my other dataframes have data on organic products, specifically? 

bls_df = pd.read_csv('cleaned_bls_cpi_data.csv')
costco_df = pd.read_csv('cleaned_costco_data.csv')

In [13]:
bls_df.head()

product  unadjusted_percent_change_2024  \
0                    All items                             2.9   
1                         Food                             2.5   
2                 Food at home                             1.8   
3  Cereals and bakery products                             0.8   
4  Cereals and cereal products                             1.7   

   seasonally_adjusted_effect_2024  unadjusted_effect_2024  
0                              NaN                     NaN  
1                            0.042                   0.340  
2                            0.026                   0.145  
3                            0.012                   0.008  
4                            0.004                   0.005

In [ ]:
# Any organic products in this table? 

bls_organic_filtered_df = bls_df[bls_df['product'].str.contains('organic', case=False, na=False)]

bls_organic_filtered_df.head()

# Apparently not. 

Empty DataFrame
Columns: [product, unadjusted_percent_change_2024, seasonally_adjusted_effect_2024, unadjusted_effect_2024]
Index: []

In [15]:
costco_df.head()

sub_category   price  \
0  Bakery & Desserts   56.99   
1  Bakery & Desserts  159.99   
2  Bakery & Desserts   44.99   
3  Bakery & Desserts   39.99   
4  Bakery & Desserts   59.99   

                                               title  
0  David’s Cookies Mile High Peanut Butter Cake, ...  
1  The Cake Bake Shop 8" Round Carrot Cake (16-22...  
2  St Michel Madeleine, Classic French Sponge Cak...  
3  David's Cookies Butter Pecan Meltaways 32 oz, ...  
4  David’s Cookies Premier Chocolate Cake, 7.2 lb...

In [ ]:
# Any organic products in this table?

costco_organic_filtered_df = costco_df[
    costco_df['sub_category'].str.contains('organic', case=False, na=False) | 
    costco_df['title'].str.contains('organic', case=False, na=False)
]

costco_organic_filtered_df.head()

# Okay, so there ARE organic products in the costco_df...

sub_category  price  \
35  Beverages & Water   9.99   
37  Beverages & Water  21.99   
46  Beverages & Water  12.99   
52  Beverages & Water  21.99   
57  Beverages & Water  21.99   

                                                title  
35  Kirkland Signature, Organic Almond Beverage, V...  
37  Kirkland Signature, Organic Reduced Fat Chocol...  
46  SO Delicious, Organic Coconut Milk, 32 oz, 6-C...  
52  Kirkland Signature, Organic Coconut Water, 33....  
57        Horizon, Organic Whole Milk, 8 oz, 18-Count

In [18]:
# What % of the products in the costco_df are organic?

organic_pct = costco_df['title'].str.contains('organic', case=False, na=False).sum() / len(costco_df['title']) * 100

print(organic_pct)

8.487084870848708


### ii. A deep dive on the specialty column

In [ ]:
usda_df['specialty'].unique()

array(['NO ', 'YES', nan], dtype=object)

In [22]:
# What % of specialty is Nan?

nan_percentage = usda_df['specialty'].isna().sum() / len(usda_df['specialty']) * 100
print(nan_percentage)

58.29050461583623


In [26]:
# What % of specialty is 'YES'

specialty_pct = (usda_df['specialty'] == 'YES').sum() / len(usda_df['specialty']) * 100
print(specialty_pct)

7.078459441625943


In [25]:
# What kinds of products are listed as specialty? 

specialty_df = usda_df[usda_df['specialty'] == 'YES']

specialty_df.head()

issuing_office report_date  program  level_1 level_2 organic specialty  \
52  DES MOINES, IA  2024-08-30  POULTRY  CHICKEN   FRESH      NO       YES   
53  DES MOINES, IA  2024-08-30  POULTRY  CHICKEN   FRESH      NO       YES   
59  DES MOINES, IA  2024-08-30  POULTRY  CHICKEN   FRESH      NO       YES   
74  DES MOINES, IA  2024-08-30  POULTRY  CHICKEN   FRESH      NO       YES   
75  DES MOINES, IA  2024-08-30  POULTRY  CHICKEN   FRESH      NO       YES   

   unit          region  weighted_average_price  price_low  price_high  
52  NaN          HAWAII                     NaN        NaN         NaN  
53  NaN          HAWAII                     NaN        NaN         NaN  
59  NaN          ALASKA                     NaN        NaN         NaN  
74  NaN  NORTHWEST U.S.                     NaN        NaN         NaN  
75  NaN  NORTHWEST U.S.                     NaN        NaN         NaN

### Takeaways: 
* I think I should filter everything marked as 'organic', which is 8.49%, from the usda_df.
* I think I should filter everything marked as 'specialty', which is 7.08%, from the usda_df.
* I think I should then drop the organic and specialty columns. 

In [ ]:
# Filtering out organic & specialty 'YES'

usda_df = usda_df[(usda_df['organic'] != 'YES') & (usda_df['specialty'] != 'YES')]

usda_df.head()

issuing_office report_date  program level_1 level_2 organic specialty unit  \
0  DES MOINES, IA  2024-08-30  POULTRY    EGGS     NaN      NO       NO   NaN   
1  DES MOINES, IA  2024-08-30  POULTRY    EGGS     NaN      NO       NO   NaN   
2  DES MOINES, IA  2024-08-30  POULTRY    EGGS     NaN      NO       NO   NaN   
3  DES MOINES, IA  2024-08-30  POULTRY    EGGS     NaN      NO       NO   NaN   
4  DES MOINES, IA  2024-08-30  POULTRY    EGGS     NaN      NO       NO   NaN   

           region  weighted_average_price  price_low  price_high  
0  NORTHEAST U.S.                     NaN        NaN         NaN  
1  NORTHEAST U.S.                     NaN        NaN         NaN  
2        NATIONAL                     NaN        NaN         NaN  
3        NATIONAL                     NaN        NaN         NaN  
4          ALASKA                     NaN        NaN         NaN

In [29]:
# Dropping now-useless organic & specialty columns

usda_df.drop(columns=['organic', 'specialty'], axis=1, inplace=True)

usda_df.head()

issuing_office report_date  program level_1 level_2 unit          region  \
0  DES MOINES, IA  2024-08-30  POULTRY    EGGS     NaN  NaN  NORTHEAST U.S.   
1  DES MOINES, IA  2024-08-30  POULTRY    EGGS     NaN  NaN  NORTHEAST U.S.   
2  DES MOINES, IA  2024-08-30  POULTRY    EGGS     NaN  NaN        NATIONAL   
3  DES MOINES, IA  2024-08-30  POULTRY    EGGS     NaN  NaN        NATIONAL   
4  DES MOINES, IA  2024-08-30  POULTRY    EGGS     NaN  NaN          ALASKA   

   weighted_average_price  price_low  price_high  
0                     NaN        NaN         NaN  
1                     NaN        NaN         NaN  
2                     NaN        NaN         NaN  
3                     NaN        NaN         NaN  
4                     NaN        NaN         NaN

### iii. The level_2 column

In [31]:
# What % is NaN?

nan_percentage = usda_df['level_2'].isna().sum() / len(usda_df['level_2']) * 100
print(nan_percentage)

6.819216736542158


In [ ]:
# Should I filter out everythign that's Nan here?

usda_df['level_2'].unique()

# I actually think this information is necessary enough that it might be a good idea to filter out NaNs

array([nan, 'PREPARED', 'FRESH/FROZEN/PREPARED', 'FRESH', 'FROZEN',
       'FRESH/FROZEN', 'FILET MIGNON', 'TENDERLOIN', 'FLANK STEAK',
       'PORTERHOUSE STEAK', 'BNLS RIBEYE STEAK',
       'BNLS NEW YORK STRIP STEAK', 'BNLS RIBEYE ROAST',
       'BONE-IN STRIP STEAK', 'BONE-IN RIBEYE STEAK',
       'BNLS TOP SIRLOIN STEAK', 'SKIRT STEAK', 'BNLS SIRLOIN STEAK',
       'BRISKET, FLAT', 'T-BONE STEAK', 'LOIN, BNLS', 'BEEF SHORT RIBS',
       'TOP ROUND STEAK', 'BNLS TOP SIRLOIN ROAST', 'SIRLOIN STEAK',
       'MINUTE/CUBE STEAKS', 'SIRLOIN TIP STEAK', 'EYE OF ROUND STEAK',
       'STEW MEAT', 'TRI-TIP', 'TOP ROUND ROAST', 'BEEF PATTIES',
       'BOTTOM ROUND ROAST', 'SIRLOIN TIP ROAST', 'LONDON BROIL',
       'EYE OF ROUND ROAST', 'GROUND SIRLOIN', 'GROUND BEEF 90% OR MORE',
       'CHUCK/SHLDR/ARM ROAST', 'BOTTOM ROUND STEAK',
       'CHUCK/SHLDR/ARM STEAK', 'RUMP ROAST', 'GROUND ROUND', 'BRISKET',
       'BEEF BACKRIBS', 'GROUND BEEF 80-89%', 'GROUND CHUCK',
       'GROUND BEEF 70-79

In [34]:
# For example, what kind of eggs are listed that don't have any level_2 information?

eggs_check_df = usda_df[(usda_df['level_1'].str.lower() == 'eggs') & (usda_df['level_2'].isna())]

eggs_check_df.head()

issuing_office report_date  program level_1 level_2 unit          region  \
0  DES MOINES, IA  2024-08-30  POULTRY    EGGS     NaN  NaN  NORTHEAST U.S.   
1  DES MOINES, IA  2024-08-30  POULTRY    EGGS     NaN  NaN  NORTHEAST U.S.   
2  DES MOINES, IA  2024-08-30  POULTRY    EGGS     NaN  NaN        NATIONAL   
3  DES MOINES, IA  2024-08-30  POULTRY    EGGS     NaN  NaN        NATIONAL   
4  DES MOINES, IA  2024-08-30  POULTRY    EGGS     NaN  NaN          ALASKA   

   weighted_average_price  price_low  price_high  
0                     NaN        NaN         NaN  
1                     NaN        NaN         NaN  
2                     NaN        NaN         NaN  
3                     NaN        NaN         NaN  
4                     NaN        NaN         NaN

In [36]:
# I think it makes the most sense to filter out rows with level_2 NaN 

usda_df = usda_df[~usda_df['level_2'].isna()]

usda_df.head()

issuing_office report_date  program level_1                level_2 unit  \
28  DES MOINES, IA  2024-08-30  POULTRY  TURKEY               PREPARED  NaN   
29  DES MOINES, IA  2024-08-30  POULTRY  TURKEY               PREPARED  NaN   
30  DES MOINES, IA  2024-08-30  POULTRY  TURKEY  FRESH/FROZEN/PREPARED  NaN   
31  DES MOINES, IA  2024-08-30  POULTRY  TURKEY  FRESH/FROZEN/PREPARED  NaN   
32  DES MOINES, IA  2024-08-30  POULTRY  TURKEY               PREPARED  NaN   

    region  weighted_average_price  price_low  price_high  
28  HAWAII                     NaN        NaN         NaN  
29  HAWAII                     NaN        NaN         NaN  
30  HAWAII                     NaN        NaN         NaN  
31  HAWAII                     NaN        NaN         NaN  
32  ALASKA                     NaN        NaN         NaN

### iv. The unit column

In [37]:
nan_percentage = usda_df['unit'].isna().sum() / len(usda_df['unit']) * 100
print(nan_percentage)

6.678188798372739


In [38]:
# Probably helpful to only have products listed where some information is also given about how they're packaged / sold
# so that I can correctly compare to other datasets.

usda_df = usda_df[~usda_df['unit'].isna()]

usda_df.head()

issuing_office report_date  program level_1   level_2          unit  \
167  DES MOINES, IA  2024-08-30  POULTRY  TURKEY  PREPARED    16 OUNCES    
168  DES MOINES, IA  2024-08-30  POULTRY  TURKEY  PREPARED  7-10 OUNCES    
170  DES MOINES, IA  2024-08-30  POULTRY  TURKEY  PREPARED    16 OUNCES    
171  DES MOINES, IA  2024-08-30  POULTRY  TURKEY  PREPARED  7-10 OUNCES    
172  DES MOINES, IA  2024-08-30  POULTRY  TURKEY  PREPARED    16 OUNCES    

             region  weighted_average_price  price_low  price_high  
167          HAWAII                    7.12       7.12        7.12  
168          HAWAII                    4.99       4.99        4.99  
170          ALASKA                    6.96       4.99        9.49  
171          ALASKA                    3.48       3.48        3.48  
172  NORTHWEST U.S.                    7.11       6.99        7.12

### v. The region column

In [ ]:
# Do all products have region data?

nan_percentage = usda_df['region'].isna().sum() / len(usda_df['region']) * 100
print(nan_percentage)

# Answer: yes, they do. 

0.0


### vi. The weighted_average_price column

In [40]:
# Do all products have a weighted average price?

nan_percentage = usda_df['weighted_average_price'].isna().sum() / len(usda_df['weighted_average_price']) * 100
print(nan_percentage)

1.6472469972746298e-05


In [41]:
# Think it makes sense to get rid of anything that doesn't have a weighted average price, as that's what i'll
# probably most heavily rely on for analysis... 

usda_df = usda_df[~usda_df['weighted_average_price'].isna()]

usda_df.head()

issuing_office report_date  program level_1   level_2          unit  \
167  DES MOINES, IA  2024-08-30  POULTRY  TURKEY  PREPARED    16 OUNCES    
168  DES MOINES, IA  2024-08-30  POULTRY  TURKEY  PREPARED  7-10 OUNCES    
170  DES MOINES, IA  2024-08-30  POULTRY  TURKEY  PREPARED    16 OUNCES    
171  DES MOINES, IA  2024-08-30  POULTRY  TURKEY  PREPARED  7-10 OUNCES    
172  DES MOINES, IA  2024-08-30  POULTRY  TURKEY  PREPARED    16 OUNCES    

             region  weighted_average_price  price_low  price_high  
167          HAWAII                    7.12       7.12        7.12  
168          HAWAII                    4.99       4.99        4.99  
170          ALASKA                    6.96       4.99        9.49  
171          ALASKA                    3.48       3.48        3.48  
172  NORTHWEST U.S.                    7.11       6.99        7.12

In [ ]:
# Do either the price_low or price_high columns have any NaNs? 

nan_low_pct = usda_df['price_low'].isna().sum() / len(usda_df['price_low']) * 100
print(nan_low_pct)

nan_high_pct = usda_df['price_high'].isna().sum() / len(usda_df['price_high']) * 100
print(nan_high_pct)

# Seems weird that those are exactly the same... 

24.299911674601457
24.299911674601457


In [47]:
price_check_df = usda_df[(usda_df['price_low'].isna())]

price_check_df.head(100)

# It appears that the reason the Nan %s for price_low and price_high are exactly the same is that if a product doesn't
# have data for one, it won't have it for the other, either. 

issuing_office report_date  program  level_1   level_2          unit  \
260  DES MOINES, IA  2024-08-30  POULTRY   TURKEY  PREPARED    16 OUNCES    
261  DES MOINES, IA  2024-08-30  POULTRY   TURKEY  PREPARED  7-10 OUNCES    
262  DES MOINES, IA  2024-08-30  POULTRY   TURKEY  PREPARED  7-10 OUNCES    
263  DES MOINES, IA  2024-08-30  POULTRY   TURKEY  PREPARED                 
264  DES MOINES, IA  2024-08-30  POULTRY   TURKEY  PREPARED                 
..              ...         ...      ...      ...       ...           ...   
474  DES MOINES, IA  2024-08-30  POULTRY  CHICKEN     FRESH  TRAY REGULAR   
475  DES MOINES, IA  2024-08-30  POULTRY  CHICKEN     FRESH    TRAY VALUE   
476  DES MOINES, IA  2024-08-30  POULTRY  CHICKEN     FRESH  TRAY REGULAR   
477  DES MOINES, IA  2024-08-30  POULTRY  CHICKEN     FRESH         TRAY    
478  DES MOINES, IA  2024-08-30  POULTRY  CHICKEN     FRESH         TRAY    

           region  weighted_average_price  price_low  price_high  
260      NATIONAL                    6.94        NaN         NaN  
261      NATIONAL                    4.96        NaN         NaN  
262      NATIONAL                    4.36        NaN         NaN  
263      NATIONAL                    6.99        NaN         NaN  
264      NATIONAL                    4.29        NaN         NaN  
..            ...                     ...        ...         ...  
474  MIDWEST U.S.                    9.28        NaN         NaN  
475  MIDWEST U.S.                    3.56        NaN         NaN  
476  MIDWEST U.S.                    4.17        NaN         NaN  
477  MIDWEST U.S.                    3.99        NaN         NaN  
478  MIDWEST U.S.                    4.77        NaN         NaN  

[100 rows x 10 columns]

In [50]:
# Off-topic, but I'm seeing what appear to be empty cells in the unit column.  Does this fix it? 

usda_df = usda_df[(usda_df['unit'].str.strip() != '')]

usda_df.head()

# Appears to... 

issuing_office report_date  program level_1   level_2          unit  \
167  DES MOINES, IA  2024-08-30  POULTRY  TURKEY  PREPARED    16 OUNCES    
168  DES MOINES, IA  2024-08-30  POULTRY  TURKEY  PREPARED  7-10 OUNCES    
170  DES MOINES, IA  2024-08-30  POULTRY  TURKEY  PREPARED    16 OUNCES    
171  DES MOINES, IA  2024-08-30  POULTRY  TURKEY  PREPARED  7-10 OUNCES    
172  DES MOINES, IA  2024-08-30  POULTRY  TURKEY  PREPARED    16 OUNCES    

             region  weighted_average_price  price_low  price_high  
167          HAWAII                    7.12       7.12        7.12  
168          HAWAII                    4.99       4.99        4.99  
170          ALASKA                    6.96       4.99        9.49  
171          ALASKA                    3.48       3.48        3.48  
172  NORTHWEST U.S.                    7.11       6.99        7.12

In [51]:
# Okay... back to price_low / price_high... 

price_check_df = usda_df[(usda_df['price_low'].isna())]

price_check_df.head(100)

issuing_office report_date  program  level_1   level_2          unit  \
260  DES MOINES, IA  2024-08-30  POULTRY   TURKEY  PREPARED    16 OUNCES    
261  DES MOINES, IA  2024-08-30  POULTRY   TURKEY  PREPARED  7-10 OUNCES    
262  DES MOINES, IA  2024-08-30  POULTRY   TURKEY  PREPARED  7-10 OUNCES    
275  DES MOINES, IA  2024-08-30  POULTRY   TURKEY     FRESH         TRAY    
276  DES MOINES, IA  2024-08-30  POULTRY   TURKEY     FRESH         TRAY    
..              ...         ...      ...      ...       ...           ...   
499  DES MOINES, IA  2024-08-30  POULTRY  CHICKEN     FRESH    TRAY VALUE   
500  DES MOINES, IA  2024-08-30  POULTRY  CHICKEN     FRESH  TRAY REGULAR   
501  DES MOINES, IA  2024-08-30  POULTRY  CHICKEN     FRESH  TRAY REGULAR   
504  DES MOINES, IA  2024-08-30  POULTRY  CHICKEN    FROZEN       BAGGED    
505  DES MOINES, IA  2024-08-30  POULTRY  CHICKEN    FROZEN       BAGGED    

             region  weighted_average_price  price_low  price_high  
260        NATIONAL                    6.94        NaN         NaN  
261        NATIONAL                    4.96        NaN         NaN  
262        NATIONAL                    4.36        NaN         NaN  
275        NATIONAL                    5.99        NaN         NaN  
276        NATIONAL                    4.82        NaN         NaN  
..              ...                     ...        ...         ...  
499  SOUTHEAST U.S.                    2.19        NaN         NaN  
500  SOUTHEAST U.S.                    3.36        NaN         NaN  
501  SOUTHEAST U.S.                    2.80        NaN         NaN  
504  NORTHEAST U.S.                    2.49        NaN         NaN  
505  NORTHEAST U.S.                    3.09        NaN         NaN  

[100 rows x 10 columns]

In [53]:
# Honestly, I think a weighted average is fine data for this analysis.  I'll drop the other 2.as_integer_ratio

usda_df.drop(columns=['price_low', 'price_high'], axis=1, inplace=True)

usda_df.head()

issuing_office report_date  program level_1   level_2          unit  \
167  DES MOINES, IA  2024-08-30  POULTRY  TURKEY  PREPARED    16 OUNCES    
168  DES MOINES, IA  2024-08-30  POULTRY  TURKEY  PREPARED  7-10 OUNCES    
170  DES MOINES, IA  2024-08-30  POULTRY  TURKEY  PREPARED    16 OUNCES    
171  DES MOINES, IA  2024-08-30  POULTRY  TURKEY  PREPARED  7-10 OUNCES    
172  DES MOINES, IA  2024-08-30  POULTRY  TURKEY  PREPARED    16 OUNCES    

             region  weighted_average_price  
167          HAWAII                    7.12  
168          HAWAII                    4.99  
170          ALASKA                    6.96  
171          ALASKA                    3.48  
172  NORTHWEST U.S.                    7.11

### vii. The regions column, revisted

In [ ]:
# How many different regions are listed?

usda_df['region'].nunique()

# 9?  That's actually really manageable.  *thumbs up*

9

### viii. The report_date column, revisted

In [56]:
# What are the earliest and latest dates in this dataset, again? 

usda_df['report_date'] = pd.to_datetime(usda_df['report_date'])
earleist_date = usda_df['report_date'].min()
latest_date = usda_df['report_date'].max()

print(earleist_date)
print(latest_date)


2020-01-03 00:00:00
2024-08-30 00:00:00


In [57]:
# How many times does each date show up (i.e. are we getting consistent amounts data over time?)

date_counts = usda_df['report_date'].value_counts()

print(date_counts)

report_date
2021-01-22    59629
2021-03-12    58017
2020-02-14    56889
2020-02-07    56514
2020-02-28    56372
              ...  
2024-08-16     5731
2024-08-23     4905
2021-01-20     4900
2021-01-11     4459
2024-08-30     3920
Name: count, Length: 251, dtype: int64
